oak_restricted_kernel/src/virtio_console.rs

```diff
@@ -0,0 +1,208 @@
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+
+use crate::{
+    acpi::{Acpi, AcpiDevice, VIRTIO_MMIO},
+    mm::Translator,
+    ADDRESS_TRANSLATOR, GUEST_HOST_HEAP,
+};
+use alloc::{string::String, vec::Vec};
+use aml::{
+    resource::{MemoryRangeDescriptor, Resource},
+    AmlContext,
+};
+use anyhow::anyhow;
+use core::{
+    alloc::{Allocator, Layout},
+    ptr::NonNull,
+};
+use log::info;
+use oak_channel::{Read, Write};
+use spinning_top::Spinlock;
+use virtio_drivers::{DeviceType, Hal, MmioTransport, Transport, VirtIOConsole};
+use x86_64::{PhysAddr, VirtAddr};
+
+struct OakHal;
+
+impl Hal for OakHal {
+    fn dma_alloc(pages: usize) -> virtio_drivers::PhysAddr {
+        // virtio_drivers::PAGE_SIZE is private right now, this has been fixed but not in the latest
+        // release
+        let vaddr = VirtAddr::from_ptr(
+            GUEST_HOST_HEAP
+                .get()
+                .unwrap()
+                .allocate(Layout::from_size_align(pages * 0x1000, 0x1000).unwrap())
+                .expect("Failed to allocate memory for virtio MMIO")
+                .cast::<u8>()
+                .as_ptr(),
+        );
+        ADDRESS_TRANSLATOR
+            .get()
+            .unwrap()
+            .translate_virtual(vaddr)
+            .unwrap()
+            .as_u64() as usize
+    }
+
+    fn dma_dealloc(paddr: virtio_drivers::PhysAddr, pages: usize) -> i32 {
+        let vaddr = ADDRESS_TRANSLATOR
+            .get()
+            .unwrap()
+            .translate_physical(PhysAddr::new(paddr as u64))
+            .unwrap();
+        unsafe {
+            GUEST_HOST_HEAP.get().unwrap().deallocate(
+                NonNull::new(vaddr.as_mut_ptr()).unwrap(),
+                Layout::from_size_align(pages * 0x1000, 0x1000).unwrap(),
+            );
+        }
+
+        0
+    }
+
+    fn phys_to_virt(paddr: virtio_drivers::PhysAddr) -> virtio_drivers::VirtAddr {
+        ADDRESS_TRANSLATOR
+            .get()
+            .unwrap()
+            .translate_physical(PhysAddr::new(paddr as u64))
+            .unwrap()
+            .as_u64() as usize
+    }
+
+    fn virt_to_phys(vaddr: virtio_drivers::VirtAddr) -> virtio_drivers::PhysAddr {
+        ADDRESS_TRANSLATOR
+            .get()
+            .unwrap()
+            .translate_virtual(VirtAddr::new(vaddr as u64))
+            .unwrap()
+            .as_u64() as usize
+    }
+}
+
+#[repr(transparent)]
+pub struct MmioConsoleChannel<'a> {
+    inner: Spinlock<VirtIOConsole<'a, OakHal, MmioTransport>>,
+}
+
+// Safety: for now, this is safe as we don't have threads in our kernel.
```

Comment:
Please create an issue and add a TODO to improve this in future. I am worried that we forget about this if we add multi-vCPU support.

Reply:
Good point, done.